vscode-lean4/src/leanclient.ts

```diff
@@ -117,31 +118,22 @@ export class LeanClient implements Disposable {
 
         // check if the lake process will start.
         let useLake = lakeEnabled();
-        if (useLake && this.folderUri && this.folderUri.scheme === 'file') {
-            const lakefile = Uri.joinPath(this.folderUri, 'lakefile.lean').toString()
-            if (!fs.existsSync(new URL(lakefile))) {
-                useLake = false;
-            }
-        }
-
-        // This is a faster way of finding out lake doesn't work in the current workspace.
-        // The LanguageClient is much slower because it does 10 retries and everything.
-        if (useLake) {
-            // First check we have a version of lake that supports "lake serve"
-            const versionOptions = version ? ['+' + version, '--version'] : ['--version']
-            const lakeVersion = await batchExecute(executable, versionOptions, this.folderUri?.fsPath, undefined);
-            const actual = this.extractVersion(lakeVersion)
-            if (actual.compare('3.0.0') >= 0) {
-                const expectedError = 'Watchdog error: Cannot read LSP request: Stream was closed\n';
-                const serveOptions =  version ? ['+' + version, 'serve'] : ['serve'];
-                const rc = await testExecute(executable, serveOptions, this.folderUri?.fsPath, this.outputChannel, true, expectedError);
-                if (rc !== 0) {
-                    const failover = 'Lake failed, using lean instead.'
-                    console.log(failover);
-                    if (this.outputChannel) this.outputChannel.appendLine(failover);
+        if (useLake && this.folderUri) {
+            if (this.folderUri.scheme === 'file') {
+                const lakefile = Uri.joinPath(this.folderUri, 'lakefile.lean').toString()
+                if (!fs.existsSync(new URL(lakefile))) {
                     useLake = false;
+                } else {
+                    const toolchainVersion = await readLeanVersion(this.folderUri);
+                    if (toolchainVersion) {
+                        const date = this.extractToolchainDate(toolchainVersion);
```

Comment:
This is unfortunately not a solution, since not every version is a nightly.  There version can also be `leanprover/lean4:4.0.0-m3` and you can also make custom releases like `gebner/lean4:vrpcrace` (which we sometimes use for testing).  From these version numbers you can't deduce the lake version at all.

Reply:
Ok, I've re-instated the lake version check, but I've keep some "optimizations" for explicit date checking that I can do when the toolchain version is something we can recognize, like `leanprover/lean4:nightly-2022-02-01` so we get the best of both.  Optimized code path on known versions, and fall back to slower `lean --version` when necessary.